### PR TITLE
feat/ Avellaneda strategy unit tests

### PR DESCRIPTION
**Pull request by CrimsonJacket**
**Before submitting this PR, please make sure**:

- [x] Your code builds clean without any errors or warnings
- [x] You are using approved title ("feat/", "fix/", "docs/", "refactor/")

**A description of the changes proposed in the pull request**:

### CODE REVIEW

File: test/hummingbot/strategy/avellaneda_market_making/test_avellaneda_market_making.py [hunk 1]
```diff
@@ -0,0 +1,1145 @@
...
+    def test_get_price(self):
+        # Avellaneda Strategy get_price is simply a wrapper for MarketTradingPairTuple.get_mid_price()
+        self.assertEqual(self.market_info.get_mid_price(), self.strategy.get_price())
+
+    def test_get_last_price(self):
+        # TODO: Determine if the get_last_price() function is needed in Avellaneda Strategy
+        # Note: MarketTrradingPairTuple does not have a get_last_price() function
+
+        # self.assertEqual(self.market_info.get_last_price(), self.strategy.get_last_price())
+        pass
```
Comment (aarmoa) [inline test/hummingbot/strategy/avellaneda_market_making/test_avellaneda_market_making.py:290]: Is the implementation missing? If the test is not required we should remove it.

File: test/hummingbot/strategy/avellaneda_market_making/test_avellaneda_market_making.py [hunk 2]
```diff
@@ -0,0 +1,1145 @@
...
+                                             trading_pair=self.trading_pair,
+                                             is_buy=True,
+                                             base_currency=self.trading_pair.split("-")[0],
+                                             quote_currency=self.trading_pair.split("-")[1],
+                                             price=Decimal("101.0"),
+                                             quantity=Decimal("10"))
+
+        self.simulate_place_limit_order(self.strategy, self.market_info, limit_order)
+
+        self.assertEqual(1, len(self.strategy.active_orders))
```
Comment (aarmoa) [inline test/hummingbot/strategy/avellaneda_market_making/test_avellaneda_market_making.py:328]: The validation should include checking that the active_order present is the one that was just simulated, maybe comparing the order id.

File: test/hummingbot/strategy/avellaneda_market_making/test_avellaneda_market_making.py [hunk 3]
```diff
@@ -0,0 +1,1145 @@
...
+                                             trading_pair=self.trading_pair,
+                                             is_buy=True,
+                                             base_currency=self.trading_pair.split("-")[0],
+                                             quote_currency=self.trading_pair.split("-")[1],
+                                             price=Decimal("101.0"),
+                                             quantity=Decimal("10"))
+
+        self.simulate_place_limit_order(self.strategy, self.market_info, limit_order)
+
+        self.assertEqual(1, len(self.strategy.active_buys))
```
Comment (aarmoa) [inline test/hummingbot/strategy/avellaneda_market_making/test_avellaneda_market_making.py:344]: Here the check should also verify that the order present is the expected one

File: test/hummingbot/strategy/avellaneda_market_making/test_avellaneda_market_making.py [hunk 4]
```diff
@@ -0,0 +1,1145 @@
...
+                                             trading_pair=self.trading_pair,
+                                             is_buy=False,
+                                             base_currency=self.trading_pair.split("-")[0],
+                                             quote_currency=self.trading_pair.split("-")[1],
+                                             price=Decimal("101.0"),
+                                             quantity=Decimal("0.5"))
+
+        self.simulate_place_limit_order(self.strategy, self.market_info, limit_order)
+
+        self.assertEqual(1, len(self.strategy.active_sells))
```
Comment (aarmoa) [inline test/hummingbot/strategy/avellaneda_market_making/test_avellaneda_market_making.py:360]: Here the check should also verify that the order present is the expected one

File: test/hummingbot/strategy/avellaneda_market_making/test_avellaneda_market_making.py [hunk 5]
```diff
@@ -0,0 +1,1145 @@
...
+        self.assertEqual(AvellanedaMarketMakingStrategy.OPTION_LOG_ALL, self.strategy.logging_options)
+
+        # Test setter method
+        self.strategy.logging_options = AvellanedaMarketMakingStrategy.OPTION_LOG_CREATE_ORDER
+
+        self.assertEqual(AvellanedaMarketMakingStrategy.OPTION_LOG_CREATE_ORDER, self.strategy.logging_options)
+
+    def test_order_tracker(self):
+        # TODO: replicate order_tracker property in Avellaneda strategy. Already exists in StrategyBase
+        pass
```
Comment (aarmoa) [inline test/hummingbot/strategy/avellaneda_market_making/test_avellaneda_market_making.py:372]: If the test has no implementation we should remove it

File: test/hummingbot/strategy/avellaneda_market_making/test_avellaneda_market_making.py [hunk 6]
```diff
@@ -0,0 +1,1145 @@
...
+    def test_execute_orders_proposal(self):
+        self.assertEqual(0, len(self.strategy.active_orders))
+
+        buys: List[PriceSize] = [PriceSize(price=Decimal("99"), size=Decimal("1"))]
+        sells: List[PriceSize] = [PriceSize(price=Decimal("101"), size=Decimal("1"))]
+        proposal: Proposal = Proposal(buys, sells)
+
+        self.strategy.execute_orders_proposal(proposal)
+
+        self.assertEqual(2, len(self.strategy.active_orders))
```
Comment (aarmoa) [inline test/hummingbot/strategy/avellaneda_market_making/test_avellaneda_market_making.py:383]: Here the check should also verify that the orders present are the expected ones

File: test/hummingbot/strategy/avellaneda_market_making/test_avellaneda_market_making.py [hunk 7]
```diff
@@ -0,0 +1,1145 @@
...
+
+        # Simulate close to _inventory_target_base_pct
+        self.market.set_balance("COINALPHA", 5)
+        self.market.set_balance("HBOT", 500)
+
+        q = (self.market.get_balance(self.base_asset) - self.strategy.calculate_target_inventory()) * self.strategy.q_adjustment_factor
+        vol = self.strategy.get_volatility()
+        min_spread, max_spread = self.strategy._get_min_and_max_spread()
+
+        # TODO: Test for expected_gamma = self.ira * (max_spread * (2-self.ira) / self.ira + min_spread) / (vol ** 2)
```
Comment (aarmoa) [inline test/hummingbot/strategy/avellaneda_market_making/test_avellaneda_market_making.py:512]: Is this TODO still pending resolution?

File: test/hummingbot/strategy/avellaneda_market_making/test_avellaneda_market_making.py [hunk 8]
```diff
@@ -0,0 +1,1145 @@
...
+
+        self.strategy.calculate_reserved_price_and_optimal_spread()
+
+        # Check reserved_price, optimal_ask and optimal_bid
+        self.assertAlmostEqual(expected_reserved_price, self.strategy.reserved_price, 2)
+        self.assertAlmostEqual(expected_optimal_spread, self.strategy.optimal_spread, 2)
+        self.assertAlmostEqual(expected_optimal_ask, self.strategy.optimal_ask, 1)
+        self.assertAlmostEqual(expected_optimal_bid, self.strategy.optimal_bid, 1)
+
+        # TODO: Test for different paths optimal_ask and optimal_bid. See AvellanedaMM line 631-648.
```
Comment (aarmoa) [inline test/hummingbot/strategy/avellaneda_market_making/test_avellaneda_market_making.py:542]: Is this TODO still pending resolution?

File: test/hummingbot/strategy/avellaneda_market_making/test_avellaneda_market_making.py [hunk 9]
```diff
@@ -0,0 +1,1145 @@
...
+                                           num=2) - 1
+
+        bid_level_spreads, ask_level_spreads = self.strategy._get_logspaced_level_spreads()
+        for i, spread in enumerate(bid_level_spreads):
+            self.assertAlmostEqual(expected_bid_spreads[i], spread, 1)
+
+        for i, spread in enumerate(ask_level_spreads):
+            self.assertAlmostEqual(expected_ask_spreads[i], spread, 1)
+
+        # Simulate high volatility. TODO: Find a better max_spread parameter to better illustrate bid levels
```
Comment (aarmoa) [inline test/hummingbot/strategy/avellaneda_market_making/test_avellaneda_market_making.py:605]: Is this TODO still pending?

File: test/hummingbot/strategy/avellaneda_market_making/test_avellaneda_market_making.py [hunk 10]
```diff
@@ -0,0 +1,1145 @@
...
+            ask_price = self.market.quantize_order_price(self.trading_pair,
+                                                         self.strategy.optimal_ask + Decimal(str(ask_level_spreads[level])))
+
+            expected_buys.append(PriceSize(bid_price, order_amount))
+            expected_sells.append(PriceSize(ask_price, order_amount))
+
+        expected_proposal: Proposal = Proposal(expected_buys, expected_sells)
+        self.assertEqual(str(expected_proposal), str(self.strategy.create_base_proposal()))
+
+        pass
```
Comment (aarmoa) [inline test/hummingbot/strategy/avellaneda_market_making/test_avellaneda_market_making.py:757]: It seems like this `pass` is no longer required

**Review by aarmoa — CHANGES_REQUESTED** [review]

File: test/hummingbot/strategy/avellaneda_market_making/test_avellaneda_market_making.py (hunk 2 above)
Comment (CrimsonJacket) [reply test/hummingbot/strategy/avellaneda_market_making/test_avellaneda_market_making.py:328]: The `BacktestMarket` class creates a limit order as follows:
`<buy/sell>://<trading_pair>/<random_uuid>`
But yes I agree that I should include a better assertion check for this.

**Review by aarmoa — DISMISSED** [review]
I agree with the tests added for Avellaneda strategy

File: test/hummingbot/strategy/avellaneda_market_making/test_avellaneda_market_making.py (hunk 9 above)
Comment (CrimsonJacket) [reply test/hummingbot/strategy/avellaneda_market_making/test_avellaneda_market_making.py:613]: See comment [here](https://github.com/CoinAlpha/hummingbot/pull/3483#discussion_r655374449).

File: test/hummingbot/strategy/avellaneda_market_making/test_avellaneda_market_making.py (hunk 8 above)
Comment (CrimsonJacket) [reply test/hummingbot/strategy/avellaneda_market_making/test_avellaneda_market_making.py:550]: See comment [here](https://github.com/CoinAlpha/hummingbot/pull/3483#discussion_r655374449).

File: hummingbot/strategy/avellaneda_market_making/avellaneda_market_making.pyx [hunk 11]
```diff
@@ -381,6 +381,10 @@ cdef class AvellanedaMarketMakingStrategy(StrategyBase):
     def max_order_age(self):
         return self._max_order_age
 
+    @max_order_age.setter
```
Comment (aarmoa) [inline hummingbot/strategy/avellaneda_market_making/avellaneda_market_making.pyx:384]: This seems as duplicated code (check line 276)

**Review by aarmoa — CHANGES_REQUESTED** [review]

**Review by aarmoa — APPROVED** [review]
The new tests for Avellaneda strategy look good to me

**Pull request merged**